### PR TITLE
Rules S3887, S2386: Add support for records and record structs

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/MutableFieldsShouldNotBePublicStatic.CSharp9.cs [hunk 1]
```diff
@@ -35,10 +36,16 @@ public record WhenReadonlyAndInitializedToImmutable
...
 
-    public record WhenNonReadonlyAlwaysReportPositional(string Property)
+    public record WhenNonReadonlyAlwaysReportPositional(string Property, int Value)
     {
-        public static ISet<string> iSetInitializaedWithImmutableSet = ImmutableHashSet.Create("a", "b"); // FN
-        public static IList<string> iListInitializaedWithImmutableArray = ImmutableArray.Create("a", "b"); // FN
-        public static IList<string> iListInitializaedWithImmutableList = ImmutableList.Create("a", "b"); // FN
+        // On positional records the analyzer gets called twice for the same SyntaxNode and so two similar diagnostics are reported
+        // Csaba already created an issue on roslyn for that https://github.com/dotnet/roslyn/issues/53136
+        // The issue is already resolved and so its expected that the "FP Duplicate" diagnostics will fail in some future Roslyn release (VS 17.2)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/MutableFieldsShouldNotBePublicStatic.CSharp9.cs:43]: Let's try to fix this as discussed

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Looks good so far, I'm requesting changes to workaround the duplicate reporting

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace SonarAnalyzer.Extensions
+{
+    internal static class SyntaxNodeAnalysisContextExtensions
+    {
+        /// <summary>
+        /// Roslyn invokes the analzyer twice for positional records.
+        /// The first invocation is for the class declaration and the second for the ctor represented by the positional parameter list.
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs:30]: Can be on a single line. Doesn't violate S103

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        /// <returns>
+        /// Returns <see langword="true"/> for the invocation on the class declaration and
+        /// <see langword="false"/> for the ctor invocation.
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs:34]: Can be on a single line. Doesn't violate S103

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        /// </returns>
+        /// <example>
+        /// record R(int i);
+        /// </example>
+        /// <seealso href="https://github.com/dotnet/roslyn/issues/50989"/>
+        internal static bool IsRedundantPositionalRecordContext(this SyntaxNodeAnalysisContext context)
+            => context.ContainingSymbol.Kind == SymbolKind.Method;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs:41]: `=>` formatting to match coding style. The `=>` should be at the end of the declaration
```suggestion
        internal static bool IsRedundantPositionalRecordContext(this SyntaxNodeAnalysisContext context) => 
            context.ContainingSymbol.Kind == SymbolKind.Method;
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Minor nitpicks

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs]: Done in 8baf21114883b150a33afa5b80e6a2d742192673

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs (hunk 3 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeAnalysisContextExtensions.cs]: Done in 8baf21114883b150a33afa5b80e6a2d742192673

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM 🎉

**Pull request merged**